### ISSUE TITLE
Wrong FPS reading in FG

### ISSUE THREAD

**Issue by matusbielik** [ResetXPDR/MSFS2020_AutoFPS#24]
Hello,

the program incorrectly reads FPS data from the MSFS, see the screenshot. When the game is in BG, the reading is correct. In FG the program reports way higher FPS, so all the time the max TLOD etc. is set, this bottlenecks on main thread and the framerate stays below desired target (in the screenshot the target fps is set to 75, to give it some margin, but I probably need to set it way higher.

Refresh rate of my monitor is 120Hz, refresh rate in Windows 10 is set to 60Hz. The game runs with 100% VSync, Quality DLSS, but that is probably not relevant, at ~60fps the GPU is chilling. The problem is incorrect reading, I probably have to set it to twice the target value?

Relevant PC specs: Ryzen 7 5800x3D & RTX 4090.

Can this be solved or am I doing something wrong? What other info can I provide?

Thanks for the reply and for the very promising tool!

![autofps_wrong_frps](https://github.com/ResetXPDR/MSFS2020_AutoFPS/assets/3437202/e42dd27d-4f98-4dd5-b474-0c8bb76f34f0)

**Comment by matusbielik** [comment]
So I just read the readme and the things are more clear to me now.
First, I thought that FG mean in "foreground", now I know it means frame generation.
Do I understand it correctly that AutoFPS doesn't work as desired when Nvidia DLSS (and/or VSync) is used? I have terrible tearing without VSync. Anyway, I will test it with changed settings. Might setting the target FPS higher than the actual FPS is help in this case? For example setting it to 120fps when in reality I only want 60fps.

_P.S. and little offtopic: as I think I've read somewhere that you have the same specs as I do (I'm not sure if you run 4k too though), can I kindly ask you for your graphics settings that you use to play the game, also your AutoFPS config values that enable you smooth gameplay at least on 60Hz/4k while the game still looks as it's running on a high-end machine? Thanks_

**Comment by ResetXPDR** [comment]
Quite simply, the MSFS FPS view does not show correct FG FPS. You need to use an external FPS monitoring tool to see the correct FPS, and you will find this app will read roughly the same value, noting the app value is averaged over 5 seconds, so may not be exactly the same.

I will make a screenshot shortly to show you.

**Comment by matusbielik** [comment]
Thanks for the quick response. I really believe that the MSFS debug fps shows correct values, as it really corresponds to the stuttering of the game when it shows low values and vice versa, the game is smooth when it shows green 60fps. It also corresponds very closely to values in nvidia overlay and Windows Game overlay. So it seems to me that AutoFPS is showing way higher values than is reality - I'm definitely not getting 114fps, the gameplay is not smooth. Anyway, what other monitoring tool should I try?

Or did I not understand properly and FG FPS are generated frames by Nvidia? Thanks for your patience.

**Comment by ResetXPDR** [comment]
Re: _P.S. and little offtopic: as I think I've read somewhere that you have the same specs as I do (I'm not sure if you run 4k too though), can I kindly ask you for your graphics settings that you use to play the game, also your AutoFPS config values that enable you smooth gameplay at least on 60Hz/4k while the game still looks as it's running on a high-end machine? Thanks_

In PC mode, I run all ultra settings, 4K but with secondary scaling in UserCfg.opt set to 2.0 and DLSS performance, which cancel each other out by final render but the end result looks fantastic - try it!

With AutoFPS, I am all over the place with settings as I still like to test it, but TBH even the non-expert auto settings work fine. I lock my FPS in nVidia CP to 70 and use a target FPS of 60 and it works great.

**Comment by ResetXPDR** [comment]
Re: _I really believe that the MSFS debug fps shows correct values_

I guarantee you it does not for FG enabled and I will shortly prove it. You can google it in the meantime if you wish.

**Comment by ResetXPDR** [comment]
![image](https://github.com/ResetXPDR/MSFS2020_AutoFPS/assets/156568588/311b745b-3cf3-4ffd-9281-59b8edf8bb15)

**Comment by ResetXPDR** [comment]
BTW, you can also see the effect secondary scaling of 2.0 and DLSS performance does. You end back up at 4K render res and it looks fantastic AA wise.

**Comment by ResetXPDR** [comment]
Also, any time MSFS is not the focused app, FG will become inactive, through your graphics driver not my app, and you need to be aware of that when looking at FG FPS - MSFS must have the focus.

And while it may surprise you that you are getting 114 FPS, I get this and more with my FPS uncapped but since my TV can only show 60Hz max I cap it at 70 FPS.

**Comment by matusbielik** [comment]
I don't want to argue as you probably know more about this topic, but I tried all the fps monitoring tools that I could find - see the screenshot below. Steam, Win Game overlay, [FrameView](https://www.nvidia.com/en-eu/geforce/technologies/frameview/), [FPSMonitor](https://fpsmon.com/en/), they all report 60fps, only AutoFPS reports twice the value. I don't know the program that you used in your screenshot.

![msfs_fps](https://github.com/ResetXPDR/MSFS2020_AutoFPS/assets/3437202/4bf7ae68-de6c-4605-84f8-742c53f61ab2)


Is it that these programs show "real" value that is being shown (as I have 60Hz on OS level), but your tool and AutoFPS get "all the frames", even these "FG"? Which is relevant though, when the game stutters if it's below 60? Would setting 120 target fps in AutoFPS be the solution for me?

Thanks for the tip with the graphic settings, I will try them.

**Comment by ResetXPDR** [comment]
Well that's interesting. Yours is the first case I've ever seen of this, even before my app.

Are you 100% sure FG is turned on in MSFS? Check the MSFS graphics settings screen. More importantly, if it is enabled, what happens to those FPS readings when you turn it off? Do they halve?

**Comment by ResetXPDR** [comment]
I would say from your latest screenshot that FG is actually not turned on. There is one way my app thinks it is on when it is not, as discovered with another user, and the solution in the readme is:

- If FG is being incorrectly reported as enabled by the app, the likely reason is that the FG mod had been installed and removed and the now the now greyed out MSFS FG setting is still set to on. To fix, change the DLSSG line in your UserCfg.opt file to be DLSSG 0

**Comment by matusbielik** [comment]
If I understand correctly, Nvidia DLSS is the FG, right? I get the same results (twice the value in AutoFPS both with and without FG, I loaded Manhattan (nice Empire State building, right? :) ) to put the CPU under load, see the screenshots below, with settings attached.

With DLSS settings:
![dlss_settings](https://github.com/ResetXPDR/MSFS2020_AutoFPS/assets/3437202/39e0fd2b-4a96-4ab3-a41e-5785f002af52)
![dlss](https://github.com/ResetXPDR/MSFS2020_AutoFPS/assets/3437202/3b6f50d8-296d-4135-91c4-36335fa1a468)

Without:
![no_dlss_settings](https://github.com/ResetXPDR/MSFS2020_AutoFPS/assets/3437202/9cb6323f-94bd-4b73-8883-b1af070a3393)
![no_dlss](https://github.com/ResetXPDR/MSFS2020_AutoFPS/assets/3437202/12806660-a14e-498e-bf90-96af45cea126)

The values are roughly the same, as I'm bottlenecked by the CPU. Note: I don't have any special software installed that I'd suspect could interfere here, and I'm running latest Nvidia driver 551.76 (since today, and I got doubled fps even before).

I don't have any special settings in Nvidia Control Panel set for MSFS.

Seeing your last message while I'm writing this, what FG mod do you have in mind? I certainly didn't install and/or uninstall such a thing. I'm running quite "vanilla" MSFS, I'm new to it :) I suspect the frame generation (DLSS, right?) does work indeed, you can't see it on the screenshots well but I can clearly spot the distinctive artifacts caused by DLSS when running in performance mode vs quality vs off.

I will try the settings you mentioned and also RTSS.

**Comment by ResetXPDR** [comment]
Re: Nvidia DLSS is the FG, right? 

No it is not. FG can work even without using DLSS.  Also you can see Frame Generation is greyed out on your settings windows, confirming FG is not enabled. My app is showing FG enabled because that's what it is reading from your UserOpt.cfg file. 

Given you have a 4090, therefore no need for the FG mod, the mostly likely reason for all of this is that Hardware Accelerated Graphics Scheduling turned off. This needs to be on for FG to work, then you should see the MSFS frame generation setting turn white and have a on or off option.

TLDR; turn this on, reboot and try again.

![image](https://github.com/ResetXPDR/MSFS2020_AutoFPS/assets/156568588/27a8f14b-1e54-4bfa-aaa7-ec5e2a90b51f)

**Comment by ResetXPDR** [comment]
BTW, you won't see artifacts in DLSS performance mode as long as you have secondary scaling set in UserCfg.Opt like:
![image](https://github.com/ResetXPDR/MSFS2020_AutoFPS/assets/156568588/8dcb4791-2166-4efb-9cee-0450e577ed42)

**Comment by matusbielik** [comment]
Ah, that one, thanks! Yes, that was disabled. Now all the measuring tools do show 100+ fps and MSFS debug shows ~50fps, like you said:

![new](https://github.com/ResetXPDR/MSFS2020_AutoFPS/assets/3437202/0357d4cd-fd08-434c-adc8-6a50a56d8d23)


As a seasoned gamer and sw dev, I thought this is all clear to me, but to be honest, after MSFS + FG + AutoFPS I am only more confused now :D If you can explain, I'd be grateful... What is the "real" fps then?
I mean, I want to run 60fps, I don't need more for a game of this type and I'd prefer my GPU to take it easy, for durability, electricity costs etc :) I got 4090 just for the purpose to have a GPU that is not utilized on max, if it makes sense :) Windows refresh rate is set to 60Hz in display settings especially to achieve that, and I have Vsync ON.

Why is the gpu still computing (and is it?) more than 60fps, when I don't "need" them, as I am only displaying 60 fps at 60Hz if I understand correctly? Isn't that processing power wasted? The numbers that I got with "Hardware-accelerated GPU scheduling" OFF made much more sense to me, and I wouldn't suspect anything is off before investigating what AutoFPS said.

Now when AutoFPS displays ~120fps and other tools more or less too, should my target setting AutoFPS be 60fps or 120fps or 70 as you have it?

Thanks again for you patience and for this amazing tool. I will try SecondaryScaling tip too.

**Comment by ResetXPDR** [comment]
If you are happy with 60 FPS, you can still use FG and just cap the frame rate in nVidia control panel at 60 FPS for MSFS. It'll be much easier on your GPU than trying to run non-FG 60 FPS. 

If using FG you set your FG target frame rate in the app. This is clearly covered in the readme.

And to FG or not to FG is really up to research yourself TBH. 

I've got some dev work to get back to so I will close this on the basis that I resolved your original issue and you know where to go and what to do from here onwards.

**Comment by ResetXPDR** [comment]
@matusbielik BTW, I will update the readme with your specific issue as it was not adequately covered there, so thanks for bring this to my attention.
